src/attribute_syncer.py

```diff
@@ -0,0 +1,531 @@
+            "remove": "sync_remove",
+            "warn": "manual_detected",
+        }
+        operation_type = operation_type_map.get(action.action_type, "manual_detected")
+
+        audit_params = s3_module.SyncAuditParams(
+            operation_type=operation_type,  # type: ignore[arg-type]
+            sso_user_principal_id=action.user_id,
+            sso_user_email=action.user_email,
+            group_id=action.group_id,
+            group_name=action.group_name,
+            reason=action.reason,
+            matched_attributes=action.matched_attributes,
+        )
+        audit_entry = s3_module.create_sync_audit_entry(audit_params)
+        s3_module.log_operation(audit_entry)
+    except Exception as e:
+        logger.exception(f"Failed to log audit entry for action {action.action_type}: {e}")
+
+
+@dataclass
+class SyncContext:
+    """Context for sync operations containing all required clients and config."""
+
+    identity_store_client: IdentityStoreClient
+    identity_store_id: str
+    s3_client: S3Client
+    slack_client: WebClient
+    config: SyncConfiguration
+    cache_config: cache_module.CacheConfig
+
+
+def _execute_action(
+    ctx: SyncContext,
+    action: SyncAction,
+    result: SyncOperationResult,
+) -> None:
+    """Execute a single sync action and update result statistics."""
+    if action.action_type == "add":
+        success = _execute_add_action(
+            identity_store_client=ctx.identity_store_client,
+            identity_store_id=ctx.identity_store_id,
+            action=action,
+        )
+        if success:
+            result.users_added += 1
+            _log_audit_entry(action)
+            send_notification_for_action(ctx.slack_client, action)
+        else:
+            result.errors.append(f"Failed to add {action.user_email} to {action.group_name}")
+
+    elif action.action_type == "remove":
+        success = _execute_remove_action(
+            identity_store_client=ctx.identity_store_client,
+            identity_store_id=ctx.identity_store_id,
+            action=action,
+        )
+        if success:
+            result.users_removed += 1
+            result.manual_assignments_removed += 1
+            _log_audit_entry(action)
+            send_notification_for_action(ctx.slack_client, action)
+        else:
+            result.errors.append(f"Failed to remove {action.user_email} from {action.group_name}")
+
+    elif action.action_type == "warn":
+        result.manual_assignments_detected += 1
+        _log_audit_entry(action)
+        send_notification_for_action(ctx.slack_client, action)
+
+
+def _finalize_result(result: SyncOperationResult) -> SyncOperationResult:
+    """Finalize the sync result with end time and success status."""
+    result.success = len(result.errors) == 0
+    result.end_time = datetime.now(timezone.utc)
+    result.log_completion()
+    return result
+
+
+def perform_sync(ctx: SyncContext) -> SyncOperationResult:  # noqa: PLR0912, PLR0915
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 49 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 98% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this function or break it into multiple functions.